src/buildstream/utils.py

```diff
@@ -381,9 +397,14 @@ def safe_copy(src: str, dest: str, *, copystat: bool = True, result: Optional[Fi
             raise UtilError("Failed to remove destination file '{}': {}".format(dest, e)) from e
 
     try:
-        shutil.copyfile(src, dest)
+        ret = copy_file_range(src, dest)
     except (OSError, shutil.Error) as e:
         raise UtilError("Failed to copy '{} -> {}': {}".format(src, dest, e)) from e
+    if not ret:
+        try:
+            shutil.copyfile(src, dest)
+        except (OSError, shutil.Error) as e:
+            raise UtilError("Failed to copy '{} -> {}': {}".format(src, dest, e)) from e
```

Comment:
Let's merge the two identical `except` clauses. I first wanted to suggest to keep the separate `except` clauses but use different error messages. However, if e.g. `open()` fails in `copy_file_range()`, raising a `copy_file_range()`-specific error message may be confusing.

Reply:
Fixed.